File: plugins/grok/scripts/lib/transfer.mjs
/**
 * Build a Grok handoff prompt from a Claude Code session jsonl.
 */
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

import { resolveWorkspaceRoot } from "./workspace.mjs";

const MAX_TURNS = 24;
const MAX_CHARS_PER_MSG = 4_000;
const MAX_HANDOFF_CHARS = 28_000;

function encodeProjectDir(p) {
  let abs = p;
  try {
    abs = fs.realpathSync(p);
  } catch {
    // keep
  }
  return String(abs).replace(/[^a-zA-Z0-9]+/g, "-");
}

function collectJsonl(dir, out) {
  if (!fs.existsSync(dir)) {
    return;
  }
  for (const name of fs.readdirSync(dir)) {
    if (name.endsWith(".jsonl")) {
      out.push(path.join(dir, name));
    }
  }
}

export function findLatestClaudeSession(cwd) {
  const workspace = resolveWorkspaceRoot(cwd);
  const projects = path.join(os.homedir(), ".claude", "projects");
  if (!fs.existsSync(projects)) {
    throw new Error(`No Claude projects directory at ${projects}`);
  }
  const candidates = [];
  const preferred = path.join(projects, encodeProjectDir(workspace));
  collectJsonl(preferred, candidates);
  if (candidates.length === 0) {
    const base = path.basename(workspace);
    for (const name of fs.readdirSync(projects)) {
      if (base && name.includes(base)) {
        collectJsonl(path.join(projects, name), candidates);
      }
    }
  }
  if (candidates.length === 0) {
    for (const name of fs.readdirSync(projects)) {
      collectJsonl(path.join(projects, name), candidates);
    }
  }
  candidates.sort((a, b) => {
    try {
      return fs.statSync(b).mtimeMs - fs.statSync(a).mtimeMs;
    } catch {
      return 0;
    }
  });
  if (!candidates.length) {
    throw new Error(`No Claude session jsonl found for ${workspace}`);
  }
  return candidates[0];
}

function extractContent(v) {
  if (typeof v?.content === "string") {
    return v.content;
  }
  if (Array.isArray(v?.content)) {
    return v.content
      .map((item) => item?.text || (typeof item === "string" ? item : ""))
      .filter(Boolean)
      .join("\n");
  }
  if (typeof v?.text === "string") {
    return v.text;
  }
  return "";
}

function truncateMsg(s) {
  const t = String(s).trim();
  if (t.length <= MAX_CHARS_PER_MSG) {
    return t;
  }
  return `${t.slice(0, MAX_CHARS_PER_MSG)}\n…[message truncated]…`;
}

export function buildTransferPrompt(source) {
  const raw = fs.readFileSync(source, "utf8");
  const turns = [];
  for (const line of raw.split(/\r?\n/)) {
    if (!line.trim()) {
      continue;
    }
    let v;
    try {
      v = JSON.parse(line);
    } catch {
      continue;
    }
    let role = v.role || v.type || "";
    let content = extractContent(v);
    if ((!role || !content) && v.message) {
      role = v.message.role || role;
      content = extractContent(v.message);
    }
    if (role === "human") {
      role = "user";
    }
    if (role === "ai") {
      role = "assistant";
    }
    if (role !== "user" && role !== "assistant") {
      continue;
    }
    if (!String(content).trim()) {
      continue;
    }
    turns.push([role, truncateMsg(content)]);
  }
  if (!turns.length) {
    throw new Error(`No user/assistant turns found in ${source}`);
  }
  const selected = turns.slice(-MAX_TURNS);
  let body = `# Claude → Grok session transfer

You are continuing work started in Claude Code. Below is a recent transcript excerpt.
Pick up the highest-value next step and continue until the original goal is met or blocked.
Source: \`${source}\`

## Transcript

`;
  for (const [role, text] of selected) {
    body += `### ${role}\n${text}\n\n`;
  }
  body += `## Instruction\nContinue from the latest assistant state. Prefer the smallest safe next steps.\n`;
  if (body.length > MAX_HANDOFF_CHARS) {
    body = `${body.slice(0, MAX_HANDOFF_CHARS)}\n…[transfer truncated]\n`;
  }
  return { source, prompt: body, turns: selected.length };
}

File: crates/grok-companion/src/transfer.rs
//! Transfer Claude Code session transcripts into a Grok handoff prompt.
//!
//! Claude stores sessions under `~/.claude/projects/<encoded-cwd>/*.jsonl`.
//! We extract recent user/assistant turns and start a Grok task with that context.

use std::fs;
use std::path::{Path, PathBuf};

use crate::error::{CompanionError, Result};
use crate::workspace::resolve_workspace_root;

const MAX_TURNS: usize = 24;
const MAX_CHARS_PER_MSG: usize = 4_000;
const MAX_HANDOFF_CHARS: usize = 28_000;

#[derive(Debug, Clone)]
pub struct TransferResult {
    pub source: PathBuf,
    pub prompt: String,
    pub turns: usize,
}

/// Encode a path the way Claude Code names project dirs (roughly).
fn encode_project_dir(path: &Path) -> String {
    let abs = path
        .canonicalize()
        .unwrap_or_else(|_| path.to_path_buf());
    abs.to_string_lossy()
        .chars()
        .map(|c| if c.is_ascii_alphanumeric() { c } else { '-' })
        .collect()
}

fn claude_projects_root() -> PathBuf {
    directories::UserDirs::new()
        .map(|u| u.home_dir().join(".claude/projects"))
        .unwrap_or_else(|| PathBuf::from(".claude/projects"))
}

/// Find the most recently modified `.jsonl` under the Claude project for `cwd`.
pub fn find_latest_claude_session(cwd: &Path) -> Result<PathBuf> {
    let workspace = resolve_workspace_root(cwd);
    let projects = claude_projects_root();
    if !projects.is_dir() {
        return Err(CompanionError::msg(format!(
            "No Claude projects directory at {}. Open a Claude Code session first.",
            projects.display()
        )));
    }

    let encoded = encode_project_dir(&workspace);
    let mut candidates: Vec<PathBuf> = Vec::new();

    // Exact-ish match first
    let preferred = projects.join(&encoded);
    if preferred.is_dir() {
        collect_jsonl(&preferred, &mut candidates);
    }

    // Fallback: any project dir containing workspace basename
    if candidates.is_empty() {
        let base = workspace
            .file_name()
            .and_then(|s| s.to_str())
            .unwrap_or("");
        if let Ok(entries) = fs::read_dir(&projects) {
            for ent in entries.flatten() {
                let name = ent.file_name().to_string_lossy().into_owned();
                if !base.is_empty() && name.contains(base) {
                    collect_jsonl(&ent.path(), &mut candidates);
                }
            }
        }
    }

    // Last resort: newest jsonl under all projects (can be wrong repo)
    if candidates.is_empty() {
        if let Ok(entries) = fs::read_dir(&projects) {
            for ent in entries.flatten() {
                collect_jsonl(&ent.path(), &mut candidates);
            }
        }
    }

    candidates.sort_by_key(|p| {
        fs::metadata(p)
            .and_then(|m| m.modified())
            .ok()
            .map(std::cmp::Reverse)
    });

    candidates.into_iter().next().ok_or_else(|| {
        CompanionError::msg(format!(
            "No Claude session jsonl found for workspace {} under {}.",
            workspace.display(),
            projects.display()
        ))
    })
}

fn collect_jsonl(dir: &Path, out: &mut Vec<PathBuf>) {
    let Ok(entries) = fs::read_dir(dir) else {
        return;
    };
    for ent in entries.flatten() {
        let p = ent.path();
        if p.extension().and_then(|e| e.to_str()) == Some("jsonl") {
            out.push(p);
        }
    }
}

/// Build a Grok handoff prompt from a Claude transcript jsonl.
pub fn build_transfer_prompt(source: &Path) -> Result<TransferResult> {
    let raw = fs::read_to_string(source).map_err(|e| {
        CompanionError::msg(format!("Failed to read {}: {e}", source.display()))
    })?;

    let mut turns: Vec<(String, String)> = Vec::new();
    for line in raw.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let Ok(v) = serde_json::from_str::<serde_json::Value>(line) else {
            continue;
        };
        // Common Claude Code shapes
        // Prefer nested message envelope (Claude Code jsonl)
        let msg = v.get("message").unwrap_or(&v);
        let role_raw = msg
            .get("role")
            .or_else(|| v.get("role"))
            .or_else(|| v.get("type"))
            .and_then(|r| r.as_str())
            .unwrap_or("");
        let role = match role_raw {
            "user" | "human" => "user",
            "assistant" | "ai" => "assistant",
            _ => continue,
        };
        let content = extract_content(msg);
        if content.trim().is_empty() {
            continue;
        }
        turns.push((role.into(), truncate_msg(&content)));
    }

    if turns.is_empty() {
        return Err(CompanionError::msg(format!(
            "No user/assistant turns found in {}.",
            source.display()
        )));
    }

    // Keep last N turns
    let start = turns.len().saturating_sub(MAX_TURNS);
    let selected = &turns[start..];

    let mut body = String::new();
    body.push_str("# Claude → Grok session transfer\n\n");
    body.push_str("You are continuing work started in Claude Code. Below is a recent transcript excerpt.\n");
    body.push_str("Pick up the highest-value next step and continue until the original goal is met or blocked.\n");
    body.push_str(&format!("Source: `{}`\n\n", source.display()));
    body.push_str("## Transcript\n\n");
    for (role, text) in selected {
        body.push_str(&format!("### {role}\n{text}\n\n"));
    }
    body.push_str("## Instruction\nContinue from the latest assistant state. Prefer the smallest safe next steps.\n");

    if body.len() > MAX_HANDOFF_CHARS {
        let mut end = MAX_HANDOFF_CHARS;
        while end > 0 && !body.is_char_boundary(end) {
            end -= 1;
        }
        body.truncate(end);
        body.push_str("\n…[transfer truncated]\n");
    }

    Ok(TransferResult {
        source: source.to_path_buf(),
        prompt: body,
        turns: selected.len(),
    })
}

fn extract_content(v: &serde_json::Value) -> String {
    if let Some(s) = v.get("content").and_then(|c| c.as_str()) {
        return s.to_string();
    }
    if let Some(arr) = v.get("content").and_then(|c| c.as_array()) {
        let mut parts = Vec::new();
        for item in arr {
            if let Some(t) = item.get("text").and_then(|t| t.as_str()) {
                parts.push(t.to_string());
            } else if let Some(t) = item.as_str() {
                parts.push(t.to_string());
            }
        }
        return parts.join("\n");
    }
    if let Some(s) = v.get("text").and_then(|t| t.as_str()) {
        return s.to_string();
    }
    String::new()
}

fn truncate_msg(s: &str) -> String {
    let s = s.trim();
    if s.chars().count() <= MAX_CHARS_PER_MSG {
        return s.to_string();
    }
    let truncated: String = s.chars().take(MAX_CHARS_PER_MSG).collect();
    format!("{truncated}\n…[message truncated]…")
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn build_transfer_from_jsonl() {
        let dir = tempdir().unwrap();
        let path = dir.path().join("sess.jsonl");
        fs::write(
            &path,
            r#"{"type":"user","message":{"role":"user","content":"fix the bug"}}
{"type":"assistant","message":{"role":"assistant","content":[{"type":"text","text":"I will investigate."}]}}
"#,
        )
        .unwrap();
        let t = build_transfer_prompt(&path).unwrap();
        assert!(t.prompt.contains("fix the bug"));
        assert!(t.prompt.contains("investigate"));
        assert_eq!(t.turns, 2);
    }
}
